src/components/TreebankService/TreebankService.js

```diff
@@ -16,33 +17,65 @@ const config = {
   commModes: [Destination.commModes.RECEIVE],
 };
 
-const error = (request, message) => ResponseMessage.Error(request, new Error(message));
+const error = (request, message, code) => ResponseMessage.Error(request, new Error(message), code);
+
+const redirectLink = ({ wordIds, config: c, sentenceId }) => {
+  const query = buildQueryString({ w: wordIds, config: c });
+
+  if (query === '') {
+    return sentenceId;
+  }
+
+  return `${sentenceId}?${query}`;
+};
 
 class TreebankService extends Component {
   constructor(props) {
     super(props);
 
-    this.state = { redirectTo: null };
+    this.state = {
+      arethusaLoaded: false,
+      redirectTo: null,
+    };
     this.messageHandler = this.messageHandler.bind(this);
+    this.setArethusaLoaded = this.setArethusaLoaded.bind(this);
   }
 
   componentDidMount() {
-    this.service = new MessagingService('treebank-service', new Destination({ ...config, receiverCB: this.messageHandler }));
+    this.destination = new Destination({ ...config, receiverCB: this.messageHandler });
+    this.service = new MessagingService('treebank-service', this.destination);
+
+    // eslint-disable-next-line no-undef
+    window.document.body.addEventListener('ArethusaLoaded', this.setArethusaLoaded);
   }
 
   componentWillUnmount() {
-    this.service.deregister();
+    this.destination.deregister();
+
+    // eslint-disable-next-line no-undef
+    window.document.body.removeEventListener('ArethusaLoaded', this.setArethusaLoaded);
+  }
+
+  setArethusaLoaded() {
+    this.setState({ arethusaLoaded: true });
   }
 
   messageHandler(request, responseFn) {
     const { arethusa } = this.props;
+    const { arethusaLoaded } = this.state;
     const { body } = request;
     const [name] = Object.keys(body);
 
+    if (!arethusaLoaded) {
+      responseFn(error(request, 'Arethusa is Not Loaded', ResponseMessage.errorCodes.SERVICE_UNINITIALIZED));
+    }
```

Comment:
should we return out of the method here? Otherwise the code will go on to try the request won't it?

Reply:
Good catch!